### PR TITLE
Implement user defined types

### PR DESCRIPTION
**Pull request by acquamarin**
# Description
This PR introduces the user defined types framework.

## Usage:
Cypher grammar: CREATE TYPE [TYPE_NAME] AS [TYPE]
We currently only support TYPE be a struct.
After type creation, users can use the type as if it is internal type.

### CODE REVIEW

File: src/include/binder/ddl/bound_create_type.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#pragma once
+
+#include "binder/bound_statement.h"
+
+namespace kuzu {
+namespace binder {
+
+class BoundCreateType final : public BoundStatement {
+public:
```
Comment (andyfengHKU) [inline src/include/binder/ddl/bound_create_type.h:9]: Let's use 
```
static constexpr common::ClauseType clauseType_ =  common::ClauseType::XXX
```
as our new convention

File: src/include/parser/ddl/create_type.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+#include "create_table_info.h"
+#include "parser/statement.h"
+
+namespace kuzu {
+namespace parser {
+
+class CreateType final : public Statement {
+public:
+    CreateType(const std::string& name, const std::string& dataType)
+        : Statement{common::StatementType::CREATE_TYPE}, name{name}, dataType{dataType} {}
```
Comment (andyfengHKU) [inline src/include/parser/ddl/create_type.h:12]: ditto static constexpr

File: src/include/planner/operator/ddl/logical_create_type.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+#include "planner/operator/ddl/logical_ddl.h"
+
+namespace kuzu {
+namespace planner {
+
+class LogicalCreateType : public LogicalDDL {
+public:
+    LogicalCreateType(std::string name, common::LogicalType type,
+        std::shared_ptr<binder::Expression> outputExpression)
+        : LogicalDDL{LogicalOperatorType::CREATE_TYPE, std::move(name),
```
Comment (andyfengHKU) [inline src/include/planner/operator/ddl/logical_create_type.h:12]: ditto static constexpr

File: src/include/storage/wal/wal_record.h [hunk 4]
```diff
@@ -68,6 +68,11 @@ struct WALRecord {
 
     virtual void serialize(common::Serializer& serializer) const;
     static std::unique_ptr<WALRecord> deserialize(common::Deserializer& deserializer);
+
+    template<class TARGET>
+    const TARGET& constCast() const {
```
Comment (andyfengHKU) [inline src/include/storage/wal/wal_record.h:73]: I think you can just make it cast() without const.

And for caller you could do
```
const auto& a = base.cast<>()
```

to obtain const reference

File: src/parser/transform/transform_ddl.cpp [hunk 5]
```diff
@@ -223,8 +231,12 @@ std::vector<PropertyDefinitionDDL> Transformer::transformPropertyDefinitionsDDL(
         if (property->oC_Expression()) {
             defaultValue = transformExpression(*property->oC_Expression());
         } else {
+            LogicalType type;
+            if (!LogicalType::tryConvertFromString(dataType, type)) {
+                type = *LogicalType::ANY();
```
Comment (andyfengHKU) [inline src/parser/transform/transform_ddl.cpp:236]: Shouldn't we throw?

File: src/planner/plan/append_simple.cpp [hunk 6]
```diff
@@ -36,63 +38,66 @@ namespace kuzu {
 namespace planner {
 
 void Planner::appendCreateTable(const BoundStatement& statement, LogicalPlan& plan) {
-    auto& createTable = ku_dynamic_cast<const BoundStatement&, const BoundCreateTable&>(statement);
+    auto& createTable = statement.constCast<BoundCreateTable>();
```
Comment (andyfengHKU) [inline src/planner/plan/append_simple.cpp:41]: ditto can u cast and use `const auto&`

File: src/processor/map/map_ddl.cpp [hunk 7]
```diff
@@ -17,49 +19,54 @@ using namespace kuzu::planner;
...
+    auto outSchema = logicalDDL.getSchema();
+    auto outputExpression = logicalDDL.getOutputExpression();
     return DataPos(outSchema->getExpressionPos(*outputExpression));
 }
 
 std::unique_ptr<PhysicalOperator> PlanMapper::mapCreateTable(LogicalOperator* logicalOperator) {
-    auto createTable = (LogicalCreateTable*)logicalOperator;
-    return std::make_unique<CreateTable>(createTable->getInfo()->copy(), getOutputPos(createTable),
-        getOperatorID(), createTable->getExpressionsForPrinting());
+    auto& createTable = logicalOperator->constCast<LogicalCreateTable>();
```
Comment (andyfengHKU) [inline src/processor/map/map_ddl.cpp:29]: ditto cast with const auto&

**Review by andyfengHKU — APPROVED** [review]

**Review by andyfengHKU — COMMENTED** [review]
Btw, I think we will extend user defined type to not only STRUCT soon?

**Comment by ray6080** [comment]
I'm still a bit lost on this feature. Mostly trying to understand:
1. What's our scope of base data types to be defined over? Do we have limitation on this? For now, it's just STRUCT, but I assume we will remove the restriction soon?
2. I assume users can feel free to define functions on top of defined data types. Can users also define casting functions? Suppose users define a new type `new_int` over `int64`, are we still doing implicit casting for `new_int` same as `int64`?

**Comment by acquamarin** [comment]
Yes, i think we can extend this to int64, and other types.
https://duckdb.org/docs/sql/statements/create_type.html
Somehow duckdb only allows creating a enum, struct and union type.
I would limit to struct for now.
This is a basic PR that implements the basic infrastructure for json.

File: src/parser/transform/transform_ddl.cpp (hunk 5 above)
Comment (acquamarin) [reply src/parser/transform/transform_ddl.cpp:236]: No, we should check the user defined type at binding stage.

**Pull request merged**